examples/models/llama2/source_transformation/sdpa.py

```diff
@@ -120,19 +162,78 @@ def replace_sdpa_with_simple_sdpa(module: torch.nn.Module):
     return module
 
 
+def replace_sdpa_with_qnn_sdpa(module: torch.nn.Module):
+    for name, child in module.named_children():
+        if isinstance(child, SDPA):
+            setattr(
+                module,
+                name,
+                SDPAQNN(child.kv_cache, child.dim, child.head_dim, child.n_rep),
+            )
+        else:
+            replace_sdpa_with_qnn_sdpa(child)
+    return module
+
+class KVCacheQNN(torch.nn.Module):
+    def __init__(
+        self,
+        max_batch_size: int,
+        max_seq_length: int,
+        n_heads: int,
+        head_dim: int,
+        dtype=torch.float32,
+    ):
+        super().__init__()
+        cache_shape = (max_batch_size, max_seq_length, n_heads, head_dim)
+        self.register_buffer(
+            "past_k_caches", torch.zeros(cache_shape, dtype=dtype, device="cpu"), persistent=False
+        )
+        self.register_buffer(
+            "past_v_caches", torch.zeros(cache_shape, dtype=dtype, device="cpu"), persistent=False
+        )
+
+    def update(
+        self, input_pos: torch.Tensor, k_val: torch.Tensor, v_val: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:     
+        k_out = torch.ops.aten.index_put_(self.past_k_caches,  [input_pos], k_val)
+        v_out = torch.ops.aten.index_put_(self.past_v_caches, [input_pos], v_val)
+
+        k_out = k_out.transpose(1,2)
+        v_out = v_out.transpose(1,2)
+        return k_out, v_out
+
+def replace_kv_cache_with_qnn_kv_cache(module: torch.nn.Module):
+    for name, child in module.named_children():
+        if isinstance(child, KVCache):
+            setattr(
+                module,
+                name,
+                KVCacheQNN(
+                    child.max_batch_size, 
+                    child.max_seq_length, 
+                    child.n_heads, 
+                    child.head_dim, 
+                    child.dtype
+                ),
+            )
+        else:
+            replace_kv_cache_with_qnn_kv_cache(child)
+    return module
+
+
 def replace_causal_mask(module: torch.nn.Module):
     for buffer_fqn_name, buffer in module.named_buffers():
         buffer_name = buffer_fqn_name.split(".")[-1]
         if buffer_name == "mask":
             max_seq_len = buffer.shape[-1]
             mask = torch.full(
                 (max_seq_len, max_seq_len),
-                float("-inf"),
+                float("-255"),
```

Comment:
Any specific reason we replace `inf` with `-255`?

Reply:
Acutually, we have a pass to replace inf with min or max because inf is not friendly for quantization or computation in Qnn Htp. It could result in numerical error.